File: M06/Python/T.E.U.py
def f(x):
    return -x**2 + 2*x + 1          # Função: -x² + 2x + 1

def df(x):
    return -2*x + 2                 # Derivada: -2x + 2

def encontrar_intervalos(a, b, passo=1):
    intervalos = []
    print(f"{'x':^5} | {'f(x)':^8} | {'f\'(x)':^8}")
    print("-" * 28)
    
    x_atual = a
    
    while x_atual <= b:
        x_prox = x_atual + passo
        fx_atual = f(x_atual)
        fx_prox = f(x_prox)
        dfx_atual = df(x_atual)
        dfx_prox = df(x_prox)

        # Exibir linha da tabela
        print(f"{x_atual:^5} | {fx_atual:^8.2f} | {dfx_atual:^8.2f}")

        # Se houver mudança de sinal e a derivada não mudar de sinal, armazenamos o intervalo
        if fx_atual * fx_prox < 0 and dfx_atual * dfx_prox > 0:
            intervalos.append((x_atual, x_prox))
        
        x_atual = x_prox

    return intervalos

# Teste no intervalo [-3, 3]
intervalos_raizes = encontrar_intervalos(-3, 3, 1)
print("\nOs intervalos que possuem raízes são:", intervalos_raizes)

File: M06/C++/T.E.U.cpp
#include <iostream>
#include <utility>
#include <iomanip>

using namespace std;

float f(int x) {
    return -(x * x) + (2 * x) + 1;
}

float df(int x) {
    return -(2 * x) + 2;
}

void encontrar_intervalo(int a, int b) {
    int passo = 1, contador = 0;
    pair<int, int> intervalos[2];

    int x_atual = a;

    cout << setw(5) << "x" << " | " << setw(10) << "f(x)" << " | " << setw(10) << "f'(x)" << endl;
    cout << string(33, '-') << endl;

    cout << fixed << setprecision(2);

    while (x_atual <= b) {
        int x_prox = x_atual + passo;
        float fx_atual = f(x_atual);
        float fx_prox = f(x_prox);
        float dfx_atual = df(x_atual);
        float dfx_prox = df(x_prox);

        cout << setw(5) << x_atual << " | " << setw(10) << fx_atual << " | " << setw(10) << dfx_atual << endl;

        if ((fx_atual * fx_prox < 0) && (dfx_atual * dfx_prox > 0)) {
            intervalos[contador] = make_pair(x_atual, x_prox);
            contador++;
        }

        x_atual = x_prox;
    }

    cout << "\nOs intervalos que possuem raízes são: ";
    cout << "[" << intervalos[0].first << ", " << intervalos[0].second << "]";
    cout << " e [" << intervalos[1].first << ", " << intervalos[1].second << "]" << endl;
}

int main() {
    encontrar_intervalo(-3, 3);
    return 0;
}
